lib/OpenQA/Worker/CommandHandler.pm

```diff
@@ -0,0 +1,185 @@
+
+    my $client             = $self->client;
+    my $worker             = $client->worker;
+    my $current_job        = $worker->current_job;
+    my $webui_host         = $client->webui_host;
+    my $current_webui_host = $worker->current_webui_host // 'unknown web UI host';
+
+    # handle responses to our own messages which are indicated by 'result'
+    if ($json->{result}) {
+        # handle response to update_status and filter known images
+        if (my $known_images = $json->{known_images}) {
+            if (!$current_job) {
+                log_warning("Ignoring 'known_images' from $webui_host because no job is running.");
+                return undef;
+            }
+            # FIXME: Also before refactoring upload_images() did not take any arguments. This merely
+            #        uploads images but does not filter known images as the comment in the previous code
+            #        stated. Is this wanted?
+            $current_job->upload_images($known_images);
+        }
+        return undef;
+    }
+
+    # handle commands of certain types regarding a specific job (which is supposed to be the job we're working on)
+    my $type = $json->{type};
+    if (!$type) {
+        log_warning("Ignoring WS message without type from $webui_host:\n" . pp($json));
+        return undef;
+    }
+
+    # require a job ID and ensure it matches the current job if we're already executing one
+    my $job_id = $json->{jobid};
+    if ($type ne 'info') {
+        if ($current_job) {
+            # ignore messages which do not belong to the current job
+            my $current_job_id = $current_job->id // 'another job';
+            if (!$job_id) {
+                # log a more specific warning in case of the grab_job message from a different web UI
+                if ($type eq 'grab_job' && $webui_host ne $current_webui_host) {
+                    log_warning(
+"Ignoring job assignment from $webui_host (already busy with job $current_job_id from $current_webui_host)."
+                    );
+                    return undef;
+                }
+                log_warning(
+"Ignoring WS message from $webui_host with type $type but no job ID (currently running $current_job_id for $current_webui_host):\n"
+                      . pp($json));
+                return undef;
+            }
+            if ($job_id ne $current_job_id) {
+                log_warning(
+"Ignoring WS message from $webui_host for job $job_id because that job is not running (running $current_job_id for $current_webui_host instead):\n"
+                      . pp($json));
+                return undef;
+            }
+        }
+        else {
+            # ignore messages which belong to a job
+            if ($job_id) {
+                log_warning(
+"Ignoring WS message from $webui_host with type $type and job ID $job_id (currently not executing a job):\n"
+                      . pp($json));
+                return undef;
+            }
+        }
```

Comment:
the above block unfortunately turns out ugly but I guess with the line length limit there is not much we can do except for maybe extracting a function? It is quite hard to find the "happy path" here with all the verbose error handling.